fix(NODE-4262): make startSession work without a connection

nbbeeken opened the pull request:
### Description

#### What is changing?

- Many test fix ups were needed for this change
- MongoClient is now the responsible for ClientSessions instead of the topology which only exist after connecting
  - The MongoClient took on the logic for session clean up that was in the topology.
- Since sessions are connectionless, cursors can obtain a session upon creation. This removes some logic where cursors had to perform server selection to check for session support. This can now be done just at operation time. 

#### What is the motivation for this change?

Now that connect is automatic and since server sessions are created lazily we can allow the API to make sessions before the client has been connected.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

baileympearson reviewed (changes requested):
I think I commented all the places I found - but there are a number of tests that are refactored seemingly at random while other tests in the file are not, and a number of tests that are deleted.

Non-test changes look good to me - just some small questions and suggestions 😄

nbbeeken commented:
Test changes summary:
 - `M` `test/integration/causal-consistency/causal_consistency.prose.test.js`
  - `skipLeakTests` is no longer part of the leak checker, explicit sessions should always be cleaned up, clients should always be closed. Removed usage
- `M` `test/integration/change-streams/change_stream.test.ts`
  - Removed `skipLeakTests` usage
  - destorying the stream at the `should support piping of Change Streams` cleans up the change stream, could probably use end but that's asynchronous, would require done/promisifying
  - made afterEach async/await better stack trace when I was fixing a bug
  - made beforeEach async/await that sets up mockServer, better TS, better stack trace
- `M` `test/integration/connection-monitoring-and-pooling/connection.test.js`
  - Simplified an afterEach that cleans up clients. Introduced behavior where if close throws it won't clean up both. ([see comment](https://github.com/mongodb/node-mongodb-native/pull/3286#discussion_r895843476))
  - async/await `test connect bad auth` test while debugging it
- `M` `test/integration/crud/misc_cursors.test.js`
  - remove `skipLeakTests` usage
  - async/await while debugging and Test assertion changed to check activeSessions saved on `MongoClient.s`
    - `should return implicit session to pool when client-side cursor exhausts results on initial query`
    - `should return implicit session to pool when client-side cursor exhausts results after a getMore`
- `M` `test/integration/gridfs/gridfs_stream.test.js`
  - Removed `skipLeakChecks` usage (overly strict topology requirements and redundant comments removed too)
- `M` `test/integration/load-balancers/load_balancers.spec.test.js`
  - Removed LB's bespoke timeout config. This was helpful during testing when we were implementing LB but it overrides the timeout that you provide on the CLI, which I sometimes need to set to `0` when debugging
- `M` `test/integration/node-specific/db.test.js`
  - Deleted test that doesn't offer more coverage than what we have in the connect/close behavior tests. 
- `M` `test/integration/node-specific/examples/causal_consistency.test.js`
  - Remove `skipLeakTests`
- `M` `test/integration/node-specific/mongo_client.test.ts`
  - Add `sends endSessions with noResponse set` test, ensure we don't wait for MongoDB to reply to this command
- `M` `test/integration/node-specific/operation_example.test.js`
  - Deleted `Should correctly connect with default replicasetNoOption` test because it tested against the Topology directly, and asserted nothing. By name it sounds like its intention was to cover an SDAM state which is covered in spec tests
  - Drive by assertion improvements in `Should correctly rewind and restart cursor` while debugging
- `D` `test/integration/node-specific/topology.test.js`
  - Reverting this deletion, although I doubt the merit of testing this internal behavior without more coverage that explains what the states assert about the topology. (not public API, not Spec-ed)
- `M` `test/integration/sessions/sessions.test.ts`
  - Remove `skipLeakTests`
  - `should send endSessions for multiple sessions` changed assertion to use the activeSessions on MongoClient
  - `session support detection` - new suite of tests stubs topology to support/not support sessions, proves that we will omit the session if the topology does not support it
- `M` `test/integration/shared.js`
  - TS fix
- `D` `test/integration/shared.test.js`
  - Removed, all the tests do not clean up properly and the helpers they test are being removed in the follow up "remove connect from tests" PR
- `M` `test/integration/transactions/transactions.test.ts`
  - Use client in place of topology
- `M` `test/tools/cluster_setup.sh`
  - oops!
- `M` `test/tools/common.js`
  - Unused helper can be deleted
- `M` `test/tools/runner/hooks/configuration.js`
  - Add process.pid to print out helps to quickly identify test run to send it signals (did you know if you send SIGUSR1 to node js it opens the debugger? so cool!)
- `M` `test/tools/unified-spec-runner/runner.ts`
  - Remove `skipLeakTests`
- `M` `test/unit/collection.test.ts`
  - Use async/await in hook, better TS better stack while debugging
- `A` `test/unit/cursor/abstract_cursor.test.ts`
  - Consolidated cursor behavior tests related to sessions into an abstract_cursor test
- `D` `test/unit/cursor/aggregation_cursor.test.js`
  - Tests rewritten in abstract_cursor
- `D` `test/unit/cursor/find_cursor.test.js`
  - Tests rewritten in abstract_cursor
- `M` `test/unit/sessions.test.js`
  - Utilize a client where ever topology was used

nbbeeken commented on src/operations/kill_cursors.ts:
```diff
@@ -0,0 +1,27 @@
+  }
+}
+
+defineAspects(KillCursorsOperation, [Aspect.CLOSING, Aspect.CURSOR_ITERATING]);
```
I renamed the symbol to reflect the behavior it controls

nbbeeken commented on test/integration/load-balancers/load_balancers.spec.test.js:
```diff
@@ -61,6 +61,5 @@ const SKIP = [
 ];
 
 describe('Load Balancer Unified Tests', function () {
```
It prevents the CLI timeout override from applying

nbbeeken commented on test/integration/node-specific/examples/causal_consistency.test.js:
```diff
@@ -28,21 +28,16 @@ describe('examples(causal-consistency):', function () {
 
-    test: async function () {
-      const session = client.startSession({ causalConsistency: true });
+    collection.insertOne({ darmok: 'jalad' }, { session });
```
I'm not so sure, we should look closer at this test, currently it will always pass

nbbeeken commented on test/integration/node-specific/db.test.js:
```diff
@@ -93,30 +93,6 @@ describe('Db', function () {
     }
   });
 
-  it('shouldCorrectlyReconnectWhenError', {
```
see summary

nbbeeken commented on test/tools/cluster_setup.sh:
```diff
@@ -21,7 +21,7 @@ elif [[ $1 == "sharded_cluster" ]]; then
 elif [[ $1 == "server" ]]; then
     mkdir -p $SINGLE_DIR
-    mlaunch init --dir $SINGLE_DIR --auth --username "bob" --password "pwd123" --single --setParameter enableTestCommands=1
+    mlaunch init --dir $SINGLE_DIR --single --setParameter enableTestCommands=1
```
My mistake!! 😅 reverted

nbbeeken commented on src/sessions.ts:
```diff
@@ -269,9 +268,10 @@ export class ClientSession extends TypedEventEmitter<ClientSessionEvents> {
           this.sessionPool.release(serverSession);
-          // Make sure a new serverSession never makes it on to the ClientSession
+          // Make sure a new serverSession never makes it on to this ClientSession
           Object.defineProperty(this, kServerSession, {
```
fixed!

nbbeeken commented on test/integration/shared.test.js:
```diff
@@ -1,100 +0,0 @@
-'use strict';
-const { withMonitoredClient } = require('./shared');
-const { expect } = require('chai');
```
I'll be removing the helpers in the remove connect from tests PR, and we're not writing any new tests that use the helpers in the mean time. They don't clean up correctly and are a bear 🐻 to debug

nbbeeken commented on test/integration/node-specific/operation_example.test.js:
```diff
@@ -4034,41 +4033,6 @@ describe('Operation Examples', function () {
-  /**
-   * Simple replicaset connection setup, requires a running replicaset on the correct ports
-   *
-   * @example-class Db
```
The `LINE` stuff is from 3.x when our JS doc pulled these tests out into browsable examples, we have abandoned the tooling that did that example generation. I think if we considered doing it again we would just use something that parses the AST as opposed to maintaining comments.

nbbeeken commented on test/integration/node-specific/operation_example.test.js:
```diff
@@ -4732,13 +4696,15 @@ describe('Operation Examples', function () {
           // Fetch the first object off the cursor
           cursor.next(function (err, item) {
-            test.equal(0, item.a);
+            expect(err).to.not.exist;
```
This one tests was failing and you don't get to see the real error because it's in `err` which wasn't being asserted. So I have to add err exists check and I fixed the other assertion while I was here. We've done drive by improvements like that, at the very least we need to be asserting the error does not exist.

baileympearson replied:
Sounds good 👍

nbbeeken commented on test/integration/node-specific/mongo_client.test.ts:
```diff
@@ -627,5 +627,24 @@ describe('class MongoClient', function () {
+
+      const startedEvents = [];
+      const endEvents = [];
+      client.on('commandStarted', event => startedEvents.push(event));
```
fixed!

nbbeeken commented on test/integration/node-specific/db.test.js:
```diff
@@ -93,30 +93,6 @@ describe('Db', function () {
 
-  it('shouldCorrectlyReconnectWhenError', {
-    metadata: {
-      requires: { topology: ['single', 'replicaset', 'sharded'] }
```
It demonstrates that connect can be called multiple times without issue, we have coverage for that with the #connect suite of tests that went in with auto-connect.

baileympearson replied:
That's a reasonable point, iirc I think we have a test that covers that scenario.  I'm okay removing this

nbbeeken commented on test/integration/connection-monitoring-and-pooling/connection.test.js:
```diff
@@ -162,24 +162,8 @@ describe('Connection', function () {
-      if (savedError) {
-        throw savedError;
-      }
+      if (client) await client.close();
```
I think it's something we can be okay with, close generally doesn't error but if it does, while inconvenient as it might hang the test run at the end I don't want to change this to swallow the error entirely. And then capturing the error from each and raising if one exists seems overly complex. Also if close does error the run may hang anyway so we're in a pretty bad spot if we get there.

nbbeeken commented on test/unit/cursor/abstract_cursor.test.ts:
```diff
@@ -0,0 +1,40 @@
+import { ClientSession } from '../../../src/sessions';
+import { ns } from '../../../src/utils';
+
+/** Minimal do nothing cursor to focus on testing the base cusor behavior */
```
Can you point out the coverage that isn't in here? The both used the mock server (which we shouldn't need here) and tests the existence and non-existence of the session on the cursor. The session now always exists so there isn't as much that applies.

baileympearson replied:
from slack: the last test doesn't seem to valuable and worth preserving

nbbeeken commented on test/unit/collection.test.ts:
```diff
@@ -6,8 +6,12 @@ import { cleanup, createServer, HELLO } from '../tools/mongodb-mock';
   let server = null;
-  beforeEach(() => createServer().then(_server => (server = _server)));
-  afterEach(() => cleanup());
+  beforeEach(async () => {
```
It's a fixup but it came up because I had to set a break point and see a stack trace that came from here. I would have to make this little code change again if I were to need to debug an issue here again, might as well keep it.

baileympearson replied:
okay, fine by me

nbbeeken commented on src/mongo_client.ts:
```diff
@@ -267,10 +267,15 @@ export type WithSessionCallback = (session: ClientSession) => Promise<any>;
   bsonOptions: BSONSerializeOptions;
   namespace: MongoDBNamespace;
   hasBeenClosed: boolean;
+  /**
```
Yea we should improve this, but let me expand here and you tell me what you think is good to include. For one, other driver's do not end explicit sessions for users, but we do and I'm keeping that behavior, this activeSessions set is how we do that, so it is "real" but it's not standard. Even if we didn't end explicit session for users the activeSessions set powers part of our leak checker that should tell you if you forgot to end an explicit session. And obviously there are tests (maybe many) that rely on the close cleans up all sessions functionality. 

I think I can drop the "for testing" part but any thoughts on what else would be good to capture here?

baileympearson reviewed (approved):
Assuming CI passes (it did last time), LGTM
